### PR TITLE
Fix migration

### LINKED ISSUE #3933: Running migration 0014_remove_micromasterscoursecertificate_final_grade.py on RC raises ValueError
ValueError: Cannot query "Program object (15)": Must be "Program" instance.

https://sentry.io/mit-office-of-digital-learning/micromasters/issues/522013709/

```
ValueError: Cannot query "Program object (15)": Must be "Program" instance.
(20 additional frame(s) were not displayed)
...
  File "django/db/models/sql/query.py", line 1253, in add_q
    clause, _ = self._add_q(q_object, self.used_aliases)
  File "django/db/models/sql/query.py", line 1277, in _add_q
    split_subq=split_subq,
  File "django/db/models/sql/query.py", line 1187, in build_filter
    self.check_related_objects(join_info.final_field, value, join_info.opts)
  File "django/db/models/sql/query.py", line 1050, in check_related_objects
    self.check_query_object_type(value, opts, field)
  File "django/db/models/sql/query.py", line 1033, in check_query_object_type
    (value, opts.object_name))

ValueError: Cannot query "Program object (15)": Must be "Program" instance.
```

### PR DESCRIPTION
**Pull request by annagav**
#### What are the relevant tickets?
Fix #3933

#### What's this PR do?
Fixed the migration to properly delete duplicate certificates.
Separated the migration into two parts:
 

- 0014 Delete all duplicate certificates,

- 0015 Remove the `final_grade` field

When you reverse the migrations all duplicate certificates should be created again.
When deploying to CI last migration should be `--faked`

#### How should this be manually tested?
Test with duplicate certificates for course.  
If you have already migrated grades to 0014 then either run:
`./manage.py migrate grades 0013`
`/manage.py migrate`
or 
`./manage.py migrate grades 0015 --fake`

### CODE REVIEW

**Comment by pdpinch** [comment]
@amir-qayyum-khan can you review this?

It looks like the test failure is the same tox/travis issue that's affecting all the PRs. I think we have a fix in the works, but you can test locally in the meantime. 

In earlier testing, we never tested the `users_with_dup` function. Please make sure that you hit that code.

**Comment by amir-qayyum-khan** [comment]
@annagav i was running `./manage.py migrate grades 0013`

```

Traceback (most recent call last):
  File "./manage.py", line 14, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 371, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 365, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 288, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 335, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/commands/migrate.py", line 200, in handle
    fake_initial=fake_initial,
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 121, in migrate
    state = self._migrate_all_backwards(plan, full_plan, fake=fake)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 196, in _migrate_all_backwards
    self.unapply_migration(states[migration], migration, fake=fake)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 262, in unapply_migration
    state = migration.unapply(state, schema_editor)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/migration.py", line 173, in unapply
    operation.database_backwards(self.app_label, schema_editor, from_state, to_state)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/operations/special.py", line 196, in database_backwards
    self.reverse_code(from_state.apps, schema_editor)
  File "/src/grades/migrations/0014_remove_micromasterscoursecertificate_final_grade.py", line 48, in add_final_grade
    for certificate in certificates:
  File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 272, in __iter__
    self._fetch_all()
  File "/usr/local/lib/python3.6/site-packages/nplusone/ext/django/patch.py", line 295, in fetch_all
    original_fetch_all(self)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 1179, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1064, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 100, in execute
    return super().execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/raven/contrib/django/client.py", line 123, in execute
    return real_execute(self, sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: column grades_micromasterscoursecertificate.final_grade_id does not exist
LINE 1: ...", "grades_micromasterscoursecertificate"."hash", "grades_mi...
```

**Comment by annagav** [comment]
@amir-qayyum-khan can you try running this `./manage.py migrate grades 0015 --fake` first?

**Comment by amir-qayyum-khan** [comment]
ok let me @annagav

**Comment by amir-qayyum-khan** [comment]
@annagav how can i create duplicate certifcates

**Comment by annagav** [comment]
You would need to create final grades for more than one run of the course, then calculate the hash: `hash = generate_md5('{}|{}'.format(user_id, course_run_id).encode('utf-8'))`
and create the certificate for a final grade: 
```MicromastersCourseCertificate.objects.create(
                final_grade=final_grade,
                user=final_grade.user,
                course=certificate.course,
                hash=hash
            )```

**Comment by amir-qayyum-khan** [comment]
@annagav final grade is not part of MicromastersCourseCertificate now

File: grades/migrations/0014_remove_micromasterscoursecertificate_final_grade.py [hunk 1]
```diff
@@ -20,33 +21,51 @@ def delete_duplicate_cert(apps, schema_editor):
...
             passed=True
-        ).order_by('-grade').first()
+        ).order_by('-grade')
+        certificate.final_grade = passing_final_grades.first()
         certificate.save()
+        # check if has other passing final_grades
+        for final_grade in passing_final_grades[1:]:
+            hash = generate_md5('{}|{}'.format(certificate.user_id, final_grade.course_run_id).encode('utf-8'))
+            MicromastersCourseCertificate.objects.create(
+                final_grade=final_grade,
```
Comment (amir-qayyum-khan) [inline grades/migrations/0014_remove_micromasterscoursecertificate_final_grade.py]: final grade is not part of MicromastersCourseCertificate

Comment (annagav) [reply grades/migrations/0014_remove_micromasterscoursecertificate_final_grade.py]: this is a reverse migration to add it back

**Comment by annagav** [comment]
@amir-qayyum-khan try creating duplicate certificates when you are on 0013, because 0014 is removing the duplicates. If you get errors, try to reset your branch to a commit before `Remove final_grade from MicromastersCourseCertificate` (#3920)

**Comment by amir-qayyum-khan** [comment]
it is tricky i need to revet model if i am going to use django shell

**Comment by annagav** [comment]
@amir-qayyum-khan, for creating the duplicate certificate:
`./manage.py migrate grades 0013`
then reset your branch to an earlier commit
`git reset --hard fb9b7d0f2808cad126a6ba2fd45ecbbdfb4c05e0`
then open the shell and create the certificates

**Comment by amir-qayyum-khan** [comment]
on migration 0013. I am on shell it is not allowing me to create multiple certificates.

```python
u = User.objects.get(username='myuser')
course = Course.objects.get(id=2)
f = FinalGrade.objects.get(user=u, course_run_id=5)  # for course run 4,5,6 of same course id 2
MicromastersCourseCertificate.objects.create(final_grade=f, user=u, course=course)
```

I am getting this error message
IntegrityError: duplicate key value violates unique constraint "grades_micromasterscoursecertificate_hash_key"
DETAIL:  Key (hash)=(b57a73c9e520db61b3c3def338b46e47) already exists.

**Comment by annagav** [comment]
Yes, because the hash has to be unique. In the earlier comments I wrote:

calculate the hash: hash = generate_md5('{}|{}'.format(user_id, course_run_id).encode('utf-8'))
and create the certificate for a final grade:
```
MicromastersCourseCertificate.objects.create(
                final_grade=final_grade,
                user=final_grade.user,
                course=certificate.course,
                hash=hash
            )
```

**Comment by amir-qayyum-khan** [comment]
@annagav you were right. It worked.
```python
from django.contrib.auth.models import User
from courses.models import Course
from grades.models import FinalGrade, MicromastersCourseCertificate
from micromasters.utils import generate_md5

u = User.objects.get(username='myuser')
course = Course.objects.get(id=2)
for run_id in range(4, 7):
    f = FinalGrade.objects.get(user=u, course_run_id=run_id)  # for course run 4,5,6 of same course id 2
    hash = generate_md5('{}|{}'.format(u.id, run_id).encode('utf-8'))
    MicromastersCourseCertificate.objects.create(final_grade=f, user=u, course=course, hash=hash)
```

**Pull request merged**